datafusion-examples/examples/simple_udwf.rs

```diff
@@ -0,0 +1,210 @@
+    println!("pwd: {}", std::env::current_dir().unwrap().display());
+    let csv_path = format!("datafusion/core/tests/data/cars.csv");
+    let read_options = CsvReadOptions::default().has_header(true);
+
+    ctx.register_csv("cars", &csv_path, read_options).await?;
+    Ok(ctx)
+}
+
+/// In this example we will declare a user defined window function that computes a moving average and then run it using SQL
+#[tokio::main]
+async fn main() -> Result<()> {
+    let ctx = create_context().await?;
+
+    // register the window function with DataFusion so wecan call it
+    ctx.register_udwf(my_average());
+
+    // Use SQL to run the new window function
+    let df = ctx.sql("SELECT * from cars").await?;
+    // print the results
+    df.show().await?;
+
+    // Use SQL to run the new window function
+    // `PARTITION BY car`:each distinct value of car (red, and green) should be treated separately
+    // `ORDER BY time`: within each group (greed or green) the values will be orderd by time
+    let df = ctx
+        .sql(
+            "SELECT car, \
+                      speed, \
+                      lag(speed, 1) OVER (PARTITION BY car ORDER BY time),\
+                      my_average(speed) OVER (PARTITION BY car ORDER BY time),\
+                      time \
+                      from cars",
+        )
+        .await?;
+    // print the results
+    df.show().await?;
+
+    // // ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING: Run the window functon so that each invocation only sees 5 rows: the 2 before and 2 after) using
+    // let df = ctx.sql("SELECT car, \
+    //                   speed, \
+    //                   lag(speed, 1) OVER (PARTITION BY car ORDER BY time ROWS BETWEEN 2 PRECEDING AND 2 FOLLOWING),\
+    //                   time \
+    //                   from cars").await?;
+    // // print the results
+    // df.show().await?;
+
+    // todo show how to run dataframe API as well
+
+    Ok(())
+}
+
+// TODO make a helper funciton like `crate_udf` that helps to make these signatures
+
+fn my_average() -> WindowUDF {
+    WindowUDF {
+        name: String::from("my_average"),
+        // it will take 2 arguments -- the column and the window size
+        signature: Signature::exact(vec![DataType::Int32], Volatility::Immutable),
+        return_type: Arc::new(return_type),
+        partition_evaluator: Arc::new(make_partition_evaluator),
+    }
+}
+
+/// Compute the return type of the function given the argument types
+fn return_type(arg_types: &[DataType]) -> Result<Arc<DataType>> {
+    if arg_types.len() != 1 {
+        return Err(DataFusionError::Plan(format!(
+            "my_udwf expects 1 argument, got {}: {:?}",
+            arg_types.len(),
+            arg_types
+        )));
+    }
+    Ok(Arc::new(arg_types[0].clone()))
+}
+
+/// Create a partition evaluator for this argument
+fn make_partition_evaluator() -> Result<Box<dyn PartitionEvaluator>> {
+    Ok(Box::new(MyPartitionEvaluator::new()))
+}
```

Comment:
Is it possible we could support passing scalar arguments when creating an instance of the function, similar to the built-in functions?

For example, the `lag` function takes an optional scalar value for the second argument, which is the shift offset:

https://github.com/apache/arrow-datafusion/blob/a42cc8d98b6e875c485e7e9b106d30803a32b00a/datafusion/core/src/physical_plan/windows/mod.rs#L148-L152

Reply:
I will try and figure out how to do this

Reply:
@stuartcarnie  I looked into adding the arguments. The primary issue I encountered is that a `WindowUDF` is specified in terms of structures in `datafusion-expr` ( aka it doesn't have access to `PhysicalExpr`s as those are defined in a different crate.

Here are some possible signatures we could provide. Do you have any feedback on these possibilities?

# Pass in the `Expr`s from the logical plan 
This is non ideal in my mind as the PartitionEvaluator is created during execution (where the `Expr`s are normally not around anymore)

```rust
/// Factory that creates a PartitionEvaluator for the given window function.
///
/// This function is passed its input arguments so that cases such as
/// constants can be correctly handled.
pub type PartitionEvaluatorFunctionFactory =
    Arc<dyn Fn(&[Expr]) -> Result<Box<dyn PartitionEvaluator>> + Send + Sync>;
```

# Pass in a `ArgType` enum
This is also non ideal in my mind as it seemingly artificially limits what the user defined window function can special case (why not Column's for example??)

```rust
enum ArgType {
  /// The argument was a single value
  Scalar(ScalarValue),
  /// the argument is something other than a single value
  Array
}

/// Factory that creates a PartitionEvaluator for the given window function.
///
/// This function is passed its input arguments so that cases such as
/// constants can be specially handled if desired.
pub type PartitionEvaluatorFunctionFactory =
    Arc<dyn Fn(args: Vec<ArgType>) -> Result<Box<dyn PartitionEvaluator>> + Send + Sync>;
```

Reply:
What do you think of passing the `PhysicalExpr` trait objects, like `create_built_in_window_expr`:

https://github.com/apache/arrow-datafusion/blob/a42cc8d98b6e875c485e7e9b106d30803a32b00a/datafusion/core/src/physical_plan/windows/mod.rs#L120-L125
https://github.com/stuartcarnie/arrow-datafusion/blob/a42cc8d98b6e875c485e7e9b106d30803a32b00a/datafusion/core/src/physical_plan/windows/mod.rs#L120-L125
https://github.com/alamb/arrow-datafusion/blob/a42cc8d98b6e875c485e7e9b106d30803a32b00a/datafusion/core/src/physical_plan/windows/mod.rs#L120-L125

such as:

```rust
/// Factory that creates a PartitionEvaluator for the given window function.
///
/// This function is passed its input arguments so that cases such as
/// constants can be correctly handled.
pub type PartitionEvaluatorFunctionFactory =
    Arc<dyn Fn(&[Arc<dyn PhysicalExpr>]) -> Result<Box<dyn PartitionEvaluator>> + Send + Sync>;
```

Reply:
> What do you think of passing the PhysicalExpr trait objects? Example:


I think this would be ideal, the problem is that `PhysicalExpr` is defined in `datafusion-physical-expr` which is not a dependency of `datafusion-expr` (the dependency goes the other way): https://github.com/apache/arrow-datafusion/blob/6194d588d5c3e9f202a31a0c524f63e6fb08d040/datafusion/physical-expr/Cargo.toml#L54

Thus, since `WindowUDF` is defined in `datafusion-expr` it can't depend on `PhysicalExpr`

datafusion_expr: https://github.com/apache/arrow-datafusion/blob/6194d588d5c3e9f202a31a0c524f63e6fb08d040/datafusion/expr/Cargo.toml#L37

Reply:
Ah yes, of course! 

I'd suggest we don't hold up this work and move this problem to another PR to solve it for both user-defined aggregate and window functions.

Reply:
The actual runtime passes `ColumnarValue`: https://docs.rs/datafusion/latest/datafusion/physical_plan/enum.ColumnarValue.html

Which is either a scalar or an array

We could potentially update the signatures to accept that instead maybe (though we would have to move it to the datafusion_expr crate)